File: sdk/java/src/test/java/jp/kengos/shojiku/EnvTest.java
package jp.kengos.shojiku;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.File;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/**
 * One flag governs every SHOJIKU_* lookup.
 *
 * <p>One rather than one per variable is the reference decision the other six SDKs mirror: an
 * application that wants a hermetic configuration wants all of it off, and a per-variable set of
 * knobs is a shape nobody can keep consistent across seven languages.
 */
class EnvTest extends Fixtures {

  @Test
  void anEnabledLookupReadsTheVariable() {
    Env env = new Env(true, Map.of("SHOJIKU_TEMPLATE_ROOT", "/templates"));

    assertEquals("/templates", env.get("SHOJIKU_TEMPLATE_ROOT"));
  }

  @Test
  void aDisabledLookupBehavesExactlyAsAnUnsetVariableDoes() {
    // So calling code has no second branch to get wrong.
    Env disabled = new Env(false, Map.of("SHOJIKU_TEMPLATE_ROOT", "/templates"));

    assertNull(disabled.get("SHOJIKU_TEMPLATE_ROOT"));
    assertEquals(List.of(), disabled.paths("SHOJIKU_FONT_DIR"));
  }

  @Test
  void anUnsetOrBlankVariableIsNothingRatherThanAnEmptyString() {
    assertNull(new Env(true, Map.of()).get("SHOJIKU_TEMPLATE_ROOT"));
    assertNull(new Env(true, Map.of("SHOJIKU_TEMPLATE_ROOT", "")).get("SHOJIKU_TEMPLATE_ROOT"));
  }

  @Test
  void severalPathsRideInOneVariableTheWayEveryOtherToolSpellsIt() {
    String separator = File.pathSeparator;
    Env env =
        new Env(
            true,
            Map.of("SHOJIKU_FONT_DIR", "/a" + separator + "/b" + separator + separator + "/c"));

    assertEquals(List.of("/a", "/b", "/c"), env.paths("SHOJIKU_FONT_DIR"));
  }

  @Test
  void anUnsetPathVariableIsNoPathsAtAll() {
    assertEquals(List.of(), new Env(true, Map.of()).paths("SHOJIKU_FONT_DIR"));
  }

  @Test
  void theRealEnvironmentIsTheDefaultSource() {
    // The gate image sets this one, which is what makes it a usable probe: a Java
    // process cannot set a variable in its own environment.
    assertEquals(LIBRARY, new Env(true).get("SHOJIKU_LIBRARY"));
    assertNull(new Env(false).get("SHOJIKU_LIBRARY"));
  }

  @Test
  void oneFlagGovernsThePackDirectoriesToo() {
    // Off, and with nothing configured either: the packs come from nowhere, so a
    // template needing a font cannot render.
    ShojikuClient hermetic =
        ShojikuClient.builder().templates(TEMPLATES).library(LIBRARY).env(false).build();

    assertTrue(hermetic.generate("receipt", null).failed());
  }

  @Test
  void explicitPackDirectoriesAreWhatTheClientUses() {
    // The same client with the packs named explicitly renders — which is the other
    // half of the assertion above.
    assertTrue(client().build().generate("receipt", null).success());
  }
}

File: sdk/js/test/env.test.ts
/**
 * The environment, and the ONE flag that governs every `SHOJIKU_*` lookup.
 *
 * One flag rather than one per variable is the reference decision the other six
 * SDKs mirror: an application that wants a hermetic configuration wants all of
 * it off.
 */

import { delimiter, join } from 'node:path';
import { afterEach, describe, expect, it } from 'vitest';
import { Env } from '../src/env.js';
import { resetConfiguration } from '../src/index.js';
import { enginePath, FIXTURE_TEMPLATES, makeClient, REPO_ROOT } from './support/fixtures.js';

afterEach(resetConfiguration);

describe('Env', () => {
  it('reads a variable when lookups are on', () => {
    expect(new Env(true, { SHOJIKU_TEMPLATE_ROOT: '/from/env' }).get('SHOJIKU_TEMPLATE_ROOT')).toBe(
      '/from/env',
    );
  });

  it('reads nothing at all when they are off', () => {
    expect(
      new Env(false, { SHOJIKU_TEMPLATE_ROOT: '/from/env' }).get('SHOJIKU_TEMPLATE_ROOT'),
    ).toBeNull();
  });

  it('treats an unset and a BLANK variable the same', () => {
    expect(new Env(true, {}).get('SHOJIKU_TEMPLATE_ROOT')).toBeNull();
    expect(new Env(true, { SHOJIKU_TEMPLATE_ROOT: '' }).get('SHOJIKU_TEMPLATE_ROOT')).toBeNull();
  });

  it('splits a multi-path variable on the platform delimiter', () => {
    const env = new Env(true, { SHOJIKU_FONT_DIR: ['/a', '/b'].join(delimiter) });

    expect(env.paths('SHOJIKU_FONT_DIR')).toEqual(['/a', '/b']);
  });

  it('drops empty entries in a multi-path variable', () => {
    const env = new Env(true, { SHOJIKU_FONT_DIR: `${delimiter}/a${delimiter}` });

    expect(env.paths('SHOJIKU_FONT_DIR')).toEqual(['/a']);
  });

  it('reports no paths at all when the variable is unset or lookups are off', () => {
    expect(new Env(true, {}).paths('SHOJIKU_FONT_DIR')).toEqual([]);
    expect(new Env(false, { SHOJIKU_FONT_DIR: '/a' }).paths('SHOJIKU_FONT_DIR')).toEqual([]);
  });

  it('falls back to the real process environment when given no source', () => {
    // The gate image sets this one, which is what makes it a safe probe.
    expect(new Env(true).get('SHOJIKU_LIBRARY')).toBe(enginePath());
  });
});

describe('the template root through the environment', () => {
  it('is HONOURED when nothing else configures one', async () => {
    process.env.SHOJIKU_TEMPLATE_ROOT = FIXTURE_TEMPLATES;
    try {
      const client = makeClient({ templates: null, env: true });
      expect(client.templateRoot?.path).toBe(FIXTURE_TEMPLATES);
    } finally {
      delete process.env.SHOJIKU_TEMPLATE_ROOT;
    }
  });

  it('is BEATEN by explicit configuration', () => {
    process.env.SHOJIKU_TEMPLATE_ROOT = '/from/env';
    try {
      // What an application renders is the application's own decision — the
      // deliberate opposite of how the engine library resolves.
      expect(makeClient({ env: true }).templateRoot?.path).toBe(FIXTURE_TEMPLATES);
    } finally {
      delete process.env.SHOJIKU_TEMPLATE_ROOT;
    }
  });

  it('is DISABLED by the one env flag', () => {
    process.env.SHOJIKU_TEMPLATE_ROOT = FIXTURE_TEMPLATES;
    try {
      expect(makeClient({ templates: null, env: false }).templateRoot).toBeNull();
    } finally {
      delete process.env.SHOJIKU_TEMPLATE_ROOT;
    }
  });
});

describe('the pack directories through the environment', () => {
  it('are read from SHOJIKU_FONT_DIR and SHOJIKU_LOCALE_DIR when unconfigured', async () => {
    process.env.SHOJIKU_FONT_DIR = join(REPO_ROOT, 'packs/fonts');
    process.env.SHOJIKU_LOCALE_DIR = join(REPO_ROOT, 'packs/locale');
    try {
      const client = makeClient({ fontDirs: null, localeDirs: null, env: true });
      const result = await client.generate('receipt', { customer: { name: 'x' } });

      // The packs reached the engine, or this template could not have rendered.
      expect(result.success).toBe(true);
    } finally {
      delete process.env.SHOJIKU_FONT_DIR;
      delete process.env.SHOJIKU_LOCALE_DIR;
    }
  });
});
